src/lib/substrate/context/provider.tsx

```diff
@@ -0,0 +1,242 @@
+        payload: error
+      });
+    });
+    _api.on('disconnected', () => {
+      console.log('[substrate-context API] on:disconnected');
+      dispatch({ type: ActionType.ConnectFail });
+    });
+  } catch (error) {
+    dispatch({
+      type: ActionType.ConnectError,
+      payload: error
+    });
+  }
+};
+
+const retrieveChainInfo = async (api: ApiPromise) => {
+  const [systemChain, systemChainType] = await Promise.all([
+    api.rpc.system.chain(),
+    api.rpc.system.chainType ? api.rpc.system.chainType() : Promise.resolve(registry.createType('ChainType', 'Live'))
+  ]);
+
+  return {
+    systemChain: (systemChain || '<unknown>').toString(),
+    systemChainType
+  };
+};
+
+// /
+// Loading accounts from dev and polkadot-js extension
+const loadAccounts = async (api: ApiPromise, dispatch: Dispatch): Promise<void> => {
+  dispatch({ type: ActionType.SetKeyringLoading });
+
+  try {
+    const theExtensions = await web3Enable(APP_NAME);
+    dispatch({
+      type: ActionType.SetExtensions,
+      payload: theExtensions
+    });
+
+    const theAccounts = await web3Accounts({ ss58Format: constants.SS58_FORMAT });
+    dispatch({
+      type: ActionType.SetAccounts,
+      payload: theAccounts
+    });
+
+    // Logics to check if the connecting chain is a dev chain, coming from polkadot-js Apps
+    // ref: https://github.com/polkadot-js/apps/blob/15b8004b2791eced0dde425d5dc7231a5f86c682/packages/react-api/src/Api.tsx?_pjax=div%5Bitemtype%3D%22http%3A%2F%2Fschema.org%2FSoftwareSourceCode%22%5D%20%3E%20main#L101-L110
+    const { systemChain, systemChainType } = await retrieveChainInfo(api);
+    const isDevelopment = systemChainType.isDevelopment || systemChainType.isLocal || isTestChain(systemChain);
+
+    keyring.loadAll({ isDevelopment }, theAccounts);
+
+    dispatch({
+      type: ActionType.SetKeyringReady,
+      payload: keyring
+    });
+  } catch (error) {
+    console.error('[loadAccounts] error.message => ', error.message);
+    // TODO: workaround for now
+    if (error.message === UNABLE_TO_INITIALISE_OPTIONS_MORE_THAN_ONCE) {
+      window.location.reload();
+    } else {
+      dispatch({ type: ActionType.SetKeyringError });
+    }
+  }
+};
+
+const SubstrateStateContext = React.createContext<SubstrateStateContextInterface | undefined>(undefined);
+
+const SubstrateProvider = ({ children, socket }: SubstrateProviderProps): JSX.Element => {
+  const [lsValue, setLSValue, removeLS] = useLocalStorage<_KeyringPair | undefined>(
+    SELECTED_ACCOUNT_LOCAL_STORAGE_KEY,
+    undefined
+  );
+
+  const [state, dispatch] = React.useReducer(substrateReducer, {
+    ...initialState,
+    // Filtering props and merge with default param value
+    socket: socket ?? initialState.socket,
+    selectedAccount: lsValue
+  });
+
+  const stateRef = React.useRef(state);
+  // MEMO: inspired by https://epicreact.dev/the-latest-ref-pattern-in-react/
+  React.useLayoutEffect(() => {
+    stateRef.current = state;
+  });
```

Comment:
nit: spacing

Reply:
Actually, it was deliberate for grouping.